### PR TITLE
Fix project instruments api

### LINKED ISSUE #6983: [API] The endpoint /project/{project}/intruments/{instrument} is broken
**Describe the bug**
GET `/project/{project}/intruments/{instrument}` is returning `{"error": "not found"}`

**To Reproduce**
Go to `<host>/api/v0.0.3/projects/pumpernickel/instruments/bmi`. You should see  the error.

**What did you expect to happen?**
It should return the correct information, which looks like this
instruments_bmi

### PR DESCRIPTION
**Pull request by spell00**
## Brief summary of changes

1- Changes were made to fix the endpoints at `<host>/api/v0.0.3/projects/pumpernickel/instruments/$instrumentname` , $instrumentname being the name of an instrument in the project, e.g. aosi. 

#### Link(s) to related issue(s)

* Resolves #6983

### CODE REVIEW

**Comment by spell00** [comment]
The feature added in this endpoint is not in the specs for v0.0.3, so the changes suggested in this PR need to be added to the API v0.0.4 initiated in PR #6944, which is in progress.

**Comment by spell00** [comment]
I removed the feature I previously added, as it was blocking this PR (it needs the new version of the API to be merged before). I also think it is not pertinent, so I won't make a new PR for it. It was adding the shortnames as a feature in the endpoint `/api/v0.0.3/projects/pumpernickel/instruments/`, but they are already used as the indices of the array returned by the endpoint.

**Comment by spell00** [comment]
I need to think of a new solution, I don't think the solution I am proposing is valid. I need to find a solution that works without entering a `commentID`.

**Comment by xlecours** [comment]
instead of choosing a random CommentID, you could change the getExaminerNames function in the NDB_BVL_Instrument class so that it returns an empty array when $this->getCommentID() is empty.

**Comment by spell00** [comment]
I am not sure what I should be getting in these endpoints. Here are examples (bmi returns much better information):
projects_bmi
projects_aosi

`https://<host>/api/v0.0.3/projects/pumpernickel/instruments/radiology_review ` returns `null` with the following errors: 
` PHP Notice:  Undefined index: name in /var/www/loris/php/libraries/NDB_BVL_Instrument.class.inc on line 2569` 
and 
`PHP Notice:  Undefined index: name in /var/www/loris/php/libraries/NDB_BVL_Instrument.class.inc on line 2536`

File: modules/api/php/endpoints/project/instruments.class.inc [hunk 1]
```diff
@@ -100,14 +100,10 @@ class Instruments extends Endpoint implements \LORIS\Middleware\ETagCalculator
...
 
-        $commentIDs = $this->_project->getCandidateCommentIds(
-            $instrumentname
-        );
-
         try {
             $instrument = \NDB_BVL_Instrument::factory(
                 $instrumentname,
-                $commentIDs[0],
+                null,
```
Comment (xlecours) [inline modules/api/php/endpoints/project/instruments.class.inc:105]: send an empty string instead of changing the function signature.

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 2]
```diff
@@ -177,7 +177,7 @@ abstract class NDB_BVL_Instrument extends NDB_Page
      * instrument.
      *
      * @param string  $instrument       The name of the instrument to use
-     * @param string  $commentID        The CommentID identifying the data to load
+     * @param ?string $commentID        The CommentID identifying the data to load
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Instrument.class.inc:180]: This does not have to be changed.

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 3]
```diff
@@ -192,7 +192,7 @@ abstract class NDB_BVL_Instrument extends NDB_Page
      */
     public static function factory(
         string $instrument,
-        string $commentID = '',
+        ?string $commentID = '',
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Instrument.class.inc:195]: This does not have to be changed.

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 4]
```diff
@@ -866,16 +866,19 @@ abstract class NDB_BVL_Instrument extends NDB_Page
                 'Window Difference (+/- Days)'
             );
         }
-        $examiners = $this->_getExaminerNames();
-        $this->addSelect('Examiner', 'Examiner', $examiners);
+
+        if (isset($this->commentID)) {
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Instrument.class.inc:870]: `if (empty($this->getCommentID()) {` would work with empty string and is using the getter (which is probably there for something... or not)
But this mean that the Examiner drop down wont be shown which is not ideal. It would make more sense to have `$this->_getExaminerNames()` return an empty array when there is no commentid.

File: php/libraries/NDB_BVL_Instrument_LINST.class.inc [hunk 5]
```diff
@@ -93,16 +93,19 @@ class NDB_BVL_Instrument_LINST extends \NDB_BVL_Instrument
                 'Window Difference (+/- Days)'
             );
         }
-        $examiners = $this->_getExaminerNames();
-        $this->addSelect('Examiner', 'Examiner', $examiners);
+
+        if (isset($this->commentID)) {
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Instrument_LINST.class.inc:97]: this won't be necessary if  `_getExaminerNames` is modified as mentioned above.

File: php/libraries/NDB_Page.class.inc [hunk 6]
```diff
@@ -96,16 +96,16 @@ class NDB_Page implements RequestHandlerInterface
...
      * page or specific page type.
      *
-     * @param Module $module     The test name being accessed
-     * @param string $page       The subtest being accessed (may be an empty string)
-     * @param string $identifier The identifier for the data to load on this page
-     * @param string $commentID  The CommentID to load the data for
+     * @param Module  $module     The test name being accessed
+     * @param string  $page       The subtest being accessed (may be an empty string)
+     * @param ?string $identifier The identifier for the data to load on this page
+     * @param ?string $commentID  The CommentID to load the data for
```
Comment (xlecours) [inline php/libraries/NDB_Page.class.inc:102]: unnecessary changes. you can use empty strings.

File: php/libraries/NDB_Page.class.inc [hunk 7]
```diff
@@ -96,16 +96,16 @@ class NDB_Page implements RequestHandlerInterface
...
      */
     function __construct(
         \Module $module,
         string $page,
-        string $identifier,
-        string $commentID
+        ?string $identifier,
+        ?string $commentID
```
Comment (xlecours) [inline php/libraries/NDB_Page.class.inc:108]: unnecessary changes.

File: php/libraries/NDB_Page.class.inc (hunk 6 above)
Comment (xlecours) [reply php/libraries/NDB_Page.class.inc:102]: whist I think it would be better to allow for null instead of empty strings, changing this function signature can have a lot of side effects.

**Review by xlecours — CHANGES_REQUESTED** [review]

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 8]
```diff
@@ -968,6 +969,11 @@ abstract class NDB_BVL_Instrument extends NDB_Page
         }
 
         $candID    = $_REQUEST['candID'] ?? '';
+
+        if (empty($candID)) {
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Instrument.class.inc:973]: I think the issue was CommentID. 
You could add the check at the top of the function to return an empty array where CommentID is null.

**Review by xlecours — CHANGES_REQUESTED** [review]

File: php/libraries/NDB_BVL_Instrument.class.inc (hunk 2 above)
Comment (spell00) [reply php/libraries/NDB_BVL_Instrument.class.inc]: The `setup` functions in `NDB_BVL_Instruments` and `NDB_BVL_Instruments_LINST` accepts null values for `$commentID` and `$page`. I'm removing the changes not to break anything though, as you mentioned in another comment, but ideally shouldn't it be consistent one way or the other?

**Comment by spell00** [comment]
During Travis checks, there was an error and a failure during the integration tests only for php 7.4 only. 

The error:
```
1) UserAccountsIntegrationTest::testUserAccountsFilterClearBtn

Http\Client\Exception\NetworkException: cURL error 28: Operation timed out after 150001 milliseconds with 0 out of 0 bytes received (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)

/app/vendor/php-http/guzzle6-adapter/src/Promise.php:122
/app/vendor/php-http/guzzle6-adapter/src/Promise.php:64
/app/vendor/guzzlehttp/promises/src/Promise.php:203
/app/vendor/guzzlehttp/promises/src/Promise.php:174
/app/vendor/guzzlehttp/promises/src/RejectedPromise.php:40
/app/vendor/guzzlehttp/promises/src/TaskQueue.php:47
/app/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:104
/app/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:131
/app/vendor/guzzlehttp/promises/src/Promise.php:246
/app/vendor/guzzlehttp/promises/src/Promise.php:223
/app/vendor/guzzlehttp/promises/src/Promise.php:267
/app/vendor/guzzlehttp/promises/src/Promise.php:225
/app/vendor/guzzlehttp/promises/src/Promise.php:62
/app/vendor/php-http/guzzle6-adapter/src/Promise.php:96
/app/vendor/php-http/guzzle6-adapter/src/Client.php:52
/app/src/Http/Client.php:52
/app/php/libraries/Password.class.inc:165
/app/php/libraries/Password.class.inc:118
/app/php/libraries/Password.class.inc:204
/app/test/integrationtests/LorisIntegrationTest.class.inc:81
/app/modules/user_accounts/test/user_accountsTest.php:61

Caused by

GuzzleHttp\Exception\ConnectException: cURL error 28: Operation timed out after 150001 milliseconds with 0 out of 0 bytes received (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)
```

The failure:
There was 1 failure:

```
1) UserAccountsIntegrationTest::testPasswordMustNotEqualEmail
Failed asserting that 'LORIS\n
Candidate\n
Clinical\n
Electrophysiology\n
Genomics\n
Imaging\n
Reports\n
Tools\n
Admin\n
Site Affiliations: 1\n
Unit Tester\n
User Accounts\n
Edit User\n
\n
Password Rules\n
The password must be at least 8 characters long.\n
The password cannot be your username or email address.\n
No special characters are required but your password must be sufficiently complex to be accepted.\n
Please choose a unique password.\n
We suggest using a password manager to generate one for you.\n
Notes\n
It is recommended to use an email address as the username, for clarity and uniqueness.\n
When generating a new password, please notify the user by checking 'Send email to user' box below!\n
Add/Edit User\n
User name\n
UnitTester\n
\n
Password\n
Generate new password\n
Confirm Password\n
First name *\n
Last name *\n
Degree\n
Academic Position\n
Institution\n
Department\n
Street Address\n
City\n
State/Province\n
Country\n
FAX\n
Email address *\n
  Send email to user \n
Sites *\n
Data Coordinating Center\n
Montreal\n
Ottawa\n
Rome\n
Projects *\n
Pumpernickel\n
Examiner At:\n
DCC  MTL  OTT  ROM \n
Examiner Status\n
Radiologist:  \n
Yes\n
No\n
  Pending Approval: \n
Yes\n
No\n
  Account Request Date\n
None\n
Permissions\n
Roles\n
Behavioural QC\n
Data entry\n
Superuser - supersedes all permissions\n
Permission\n
Across all sites access candidate profiles\n
Across all sites add and certify examiners\n
Across all sites create and edit users\n
Across all sites view Dicom Archive module and pages\n
Add and certify examiners\n
Add, activate, and deactivate entries in Test Battery\n
Behavioural Quality Control: view module\n
Can access only phantom data from all sites in Imaging Browser\n
Can access only phantom data from own site in Imaging Browser\n
Can add a new issue, edit own issue, comment on all\n
Can re-assign issues, mark issues as closed, comment on all, edit issues.\n
Data Release: Grant other users view-file permissions\n
Data Release: Upload file\n
Data Release: View releases\n
Delete files in Document Repository\n
Document Repository: Upload and edit files\n
Edit Acknowledgements\n
Edit Candidate Parameters\n
Edit configuration settings\n
Edit Data Dictionary\n
Edit dates of birth\n
Edit dates of death\n
Edit exclusion flags\n
Edit help documentation\n
Edit imaging browser QC status\n
Edit MRI Efax files\n
Evaluate overall exclusionary criteria for the timepoint\n
Imaging Quality Control: view module\n
Imaging Uploader\n
Instrument Builder: Create and Edit instrument forms\n
Instrument Manager: Install new instruments via file upload\n
Instrument Manager: View module\n
Manage the genomic files\n
Media files: Browsing\n
Media files: Uploading/Downloading/Editing\n
Module Manager: access the module\n
Module Manager: edit installed modules\n
Publication - Access to module\n
Publication - Approve or reject proposed publication projects\n
Publication - Propose a project\n
Resolving conflicts\n
Reverse Send from DCC\n
Send to DCC\n
Survey Accounts: view module\n
User management\n
View Acknowledgements\n
View all-sites Electrophysiology Browser pages\n
View all-sites Imaging Browser pages\n
View and complete training\n
View and manage server processes\n
View and upload files in Document Repository\n
View Battery Manager\n
View Candidate Parameters\n
View Data Dictionary (Parameter type descriptions)\n
View Data Query Tool\n
View Genomic Browser data across all sites\n
View Genomic Browser data from own site\n
View own site Electrophysiology Browser pages\n
View own-site Imaging Browser pages\n
Violated Scans: Edit MRI protocol table\n
Violated Scans: View all-sites Violated Scans\n
Violated Scans: View Violated Scans from own site\n
Supervisors\n
Data Supervisors to Email\n
Admin account\n
Unit Tester\n
| Loris Website | GitHub |\n

Powered by LORIS © 2020. All rights reserved.\n
Created by MCIN' contains "cannot be your email".
/app/modules/user_accounts/test/user_accountsTest.php:388

```

**Comment by xlecours** [comment]
This should also solve https://github.com/aces/Loris/issues/7015

**Review by xlecours — APPROVED** [review]
